### PR TITLE
Vulkan Device Feature Enable Checks

### PR DESCRIPTION
**Pull request by mark-lunarg**
With this patch series, all Vulkan device features (with some shader-related feature exceptions) now have validation layer checks.  None of these are valid usage.  

Tests can be added after the custom-device-feature-twiddling mechanisms are in place.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -7135,8 +7145,16 @@ VKAPI_ATTR void VKAPI_CALL CmdSetDepthBias(VkCommandBuffer commandBuffer, float
     GLOBAL_CB_NODE *pCB = GetCBNode(dev_data, commandBuffer);
     if (pCB) {
         skip_call |= ValidateCmd(dev_data, pCB, CMD_SETDEPTHBIASSTATE, "vkCmdSetDepthBias()");
-        UpdateCmdBufferLastCmd(pCB, CMD_SETDEPTHBIASSTATE);
-        pCB->status |= CBSTATUS_DEPTH_BIAS_SET;
+        if ((depthBiasClamp != 0.0) && (!dev_data->enabled_features.depthBiasClamp)) {
+            skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
```
Comment (tobine) [inline layers/core_validation.cpp]: This is VALIDATION_ERROR_01482

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -3271,6 +3271,14 @@ static bool verifyPipelineCreateState(layer_data *dev_data, std::vector<PIPELINE
                                          "Invalid Pipeline CreateInfo State: pDepthStencilState is NULL when rasterization is "
                                          "enabled and subpass uses a depth/stencil attachment. %s",
                                          validation_error_map[VALIDATION_ERROR_02115]);
+
+                } else if ((pPipeline->graphicsPipelineCI.pDepthStencilState->depthBoundsTestEnable == VK_TRUE) &&
+                           (!dev_data->enabled_features.depthBounds)) {
+                    skip_call |= log_msg(
+                        dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0, __LINE__,
```
Comment (tobine) [inline layers/core_validation.cpp]: This is VALIDATION_ERROR_02118. Do we also need this check for vkCmdSetScissors(), VU VALIDATION_ERROR_01497?

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -3262,6 +3262,14 @@ static bool verifyPipelineCreateState(layer_data *dev_data, std::vector<PIPELINE
         if (pPipeline->graphicsPipelineCI.pRasterizationState->rasterizerDiscardEnable == VK_FALSE) {
             auto subpass_desc = renderPass ? &renderPass->createInfo.pSubpasses[pPipeline->graphicsPipelineCI.subpass] : nullptr;
 
+            if ((pPipeline->graphicsPipelineCI.pMultisampleState->alphaToOneEnable == VK_TRUE) &&
+                (!dev_data->enabled_features.alphaToOne)) {
+                skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT,
```
Comment (tobine) [inline layers/core_validation.cpp]: Could use VALIDATION_ERROR_01464. Also do we need to check this for vkCmdSetViewport, which is actual API for 1464?

File: layers/parameter_validation.cpp [hunk 4]
```diff
@@ -3529,8 +3529,16 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSampler(VkDevice device, const VkSamplerCre
 
     skip |= parameter_validation_vkCreateSampler(report_data, pCreateInfo, pAllocator, pSampler);
 
-    // Validation for parameters excluded from the generated validation code due to a 'noautovalidity' tag in vk.xml
     if (pCreateInfo != nullptr) {
+
+        if ((device_data->physical_device_features.samplerAnisotropy == false) && (pCreateInfo->maxAnisotropy != 1.0)) {
+            skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0, __LINE__,
```
Comment (tobine) [inline layers/parameter_validation.cpp]: Can use VALIDATION_ERROR_00818

File: layers/buffer_validation.cpp [hunk 5]
```diff
@@ -2168,11 +2175,20 @@ bool ValidateBufferUsageFlags(layer_data *device_data, BUFFER_STATE const *buffe
...
+    const debug_report_data *report_data = core_validation::GetReportData(device_data);
+
     // TODO: Add check for VALIDATION_ERROR_00658
     // TODO: Add check for VALIDATION_ERROR_00666
     // TODO: Add check for VALIDATION_ERROR_00667
     // TODO: Add check for VALIDATION_ERROR_00668
     // TODO: Add check for VALIDATION_ERROR_00669
+
+    if ((pCreateInfo->flags & VK_BUFFER_CREATE_SPARSE_BINDING_BIT) && (!GetEnabledFeatures(device_data)->sparseBinding)) {
+        skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0, __LINE__,
```
Comment (tobine) [inline layers/buffer_validation.cpp]: This is VALIDATION_ERROR_00666.

File: layers/buffer_validation.cpp [hunk 6]
```diff
@@ -667,6 +667,13 @@ bool PreCallValidateCreateImage(layer_data *device_data, const VkImageCreateInfo
                              validation_error_map[VALIDATION_ERROR_00731]);
     }
 
+    if ((pCreateInfo->flags & VK_IMAGE_CREATE_SPARSE_BINDING_BIT) && (!GetEnabledFeatures(device_data)->sparseBinding)) {
+        skip_call |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0, __LINE__,
```
Comment (tobine) [inline layers/buffer_validation.cpp]: This is VALIDATION_ERROR_02143.

Comment (mark-lunarg) [inline layers/buffer_validation.cpp]: Check.

File: layers/buffer_validation.cpp (hunk 5 above)
Comment (mark-lunarg) [inline layers/buffer_validation.cpp]: Done

File: layers/core_validation.cpp [hunk 7]
```diff
@@ -3240,6 +3240,14 @@ static bool verifyPipelineCreateState(layer_data *dev_data, std::vector<PIPELINE
                                          pPipeline->graphicsPipelineCI.pRasterizationState->lineWidth);
         }
 
+        if ((pPipeline->graphicsPipelineCI.pRasterizationState->depthClampEnable == VK_TRUE) &&
+            (!dev_data->enabled_features.depthClamp)) {
+            skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
```
Comment (mark-lunarg) [inline layers/core_validation.cpp]: 1455 is correct, but the heading should be 'Rasterization', not 'CmdSetViewport' -- no changes necessary there.

File: layers/core_validation.cpp (hunk 3 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: Done.  Same as above, the 'actual API' should read 'VkPipelineMultisampleStateCreateInfo' in the Rasterization section.

File: layers/core_validation.cpp (hunk 2 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: I believe this is not a VU -- this check is to see if depthBoundsTestEnable can be set when creating a pipeline.  2118 and 1497 are checks for other settings that rely on depthBoundsTestEnable is set.  Leaving this one alone.

File: layers/core_validation.cpp [hunk 8]
```diff
@@ -6253,6 +6253,32 @@ void set_pipeline_state(PIPELINE_STATE *pPipe) {
...
+            if (VK_TRUE == pipe_state->attachments[i].blendEnable) {
+                if ((pipe_state->attachments[i].dstAlphaBlendFactor == VK_BLEND_FACTOR_SRC1_COLOR) ||
+                    (pipe_state->attachments[i].dstAlphaBlendFactor == VK_BLEND_FACTOR_ONE_MINUS_SRC1_COLOR) ||
+                    (pipe_state->attachments[i].dstAlphaBlendFactor == VK_BLEND_FACTOR_SRC1_ALPHA) ||
+                    (pipe_state->attachments[i].dstAlphaBlendFactor == VK_BLEND_FACTOR_ONE_MINUS_SRC1_ALPHA) ||
+                    (pipe_state->attachments[i].srcAlphaBlendFactor == VK_BLEND_FACTOR_SRC1_COLOR) ||
+                    (pipe_state->attachments[i].srcAlphaBlendFactor == VK_BLEND_FACTOR_ONE_MINUS_SRC1_COLOR) ||
+                    (pipe_state->attachments[i].srcAlphaBlendFactor == VK_BLEND_FACTOR_SRC1_ALPHA) ||
+                    (pipe_state->attachments[i].srcAlphaBlendFactor == VK_BLEND_FACTOR_ONE_MINUS_SRC1_ALPHA)) {
+                    skip |=
```
Comment (mark-lunarg) [inline layers/core_validation.cpp]: No, this was bogus!  Fixed.

File: layers/core_validation.cpp (hunk 1 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: Done

File: layers/parameter_validation.cpp (hunk 4 above)
Comment (mark-lunarg) [inline layers/parameter_validation.cpp]: This is a different check -- the feature here is samplerAnisotropy which governs the setting of maxAnisotropy.  818 deals with a parameter to CreateSampler and also affects the setting of maxAnisotropy.

File: layers/parameter_validation.cpp [hunk 9]
```diff
@@ -4289,8 +4289,17 @@ VKAPI_ATTR void VKAPI_CALL CmdSetLineWidth(VkCommandBuffer commandBuffer, float
 
 VKAPI_ATTR void VKAPI_CALL CmdSetDepthBias(VkCommandBuffer commandBuffer, float depthBiasConstantFactor, float depthBiasClamp,
                                            float depthBiasSlopeFactor) {
-    layer_data *my_data = GetLayerDataPtr(get_dispatch_key(commandBuffer), layer_data_map);
-    my_data->dispatch_table.CmdSetDepthBias(commandBuffer, depthBiasConstantFactor, depthBiasClamp, depthBiasSlopeFactor);
+    layer_data *device_data = GetLayerDataPtr(get_dispatch_key(commandBuffer), layer_data_map);
+    bool skip = false;
+    if ((!device_data->physical_device_features.depthBiasClamp) && (depthBiasClamp != 0.0)) {
+        skip |= log_msg(device_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
```
Comment (mark-lunarg) [inline layers/parameter_validation.cpp]: This is a redundant check already accounted for in core_validation.  Removed.

File: layers/parameter_validation.cpp [hunk 10]
```diff
@@ -4445,6 +4445,11 @@ VKAPI_ATTR void VKAPI_CALL CmdDrawIndirect(VkCommandBuffer commandBuffer, VkBuff
     layer_data *my_data = GetLayerDataPtr(get_dispatch_key(commandBuffer), layer_data_map);
     assert(my_data != NULL);
 
+    if (!my_data->physical_device_features.multiDrawIndirect && ((count != 0) && (count != 1))) {
```
Comment (mark-lunarg) [inline layers/parameter_validation.cpp]: Done

File: layers/parameter_validation.cpp [hunk 11]
```diff
@@ -4457,7 +4462,12 @@ VKAPI_ATTR void VKAPI_CALL CmdDrawIndexedIndirect(VkCommandBuffer commandBuffer,
     bool skip = false;
     layer_data *my_data = GetLayerDataPtr(get_dispatch_key(commandBuffer), layer_data_map);
     assert(my_data != NULL);
-
+    if (!my_data->physical_device_features.multiDrawIndirect && ((count != 0) && (count != 1))) {
```
Comment (mark-lunarg) [inline layers/parameter_validation.cpp]: Done

**Comment by mark-lunarg** [comment]
Man, @tobine, you are getting downright spooky with the VU numbers.  I'm guessing you must have them memorized by now!

File: layers/buffer_validation.cpp [hunk 12]
```diff
@@ -2168,11 +2184,37 @@ bool ValidateBufferUsageFlags(layer_data *device_data, BUFFER_STATE const *buffe
 
 bool PreCallValidateCreateBuffer(layer_data *device_data, const VkBufferCreateInfo *pCreateInfo) {
     bool skip = false;
+    const debug_report_data *report_data = core_validation::GetReportData(device_data);
+
     // TODO: Add check for VALIDATION_ERROR_00658
     // TODO: Add check for VALIDATION_ERROR_00666
```
Comment (tobine) [inline layers/buffer_validation.cpp]: Can remove this TODO

**Comment by mark-lunarg** [comment]
Removed comment and pushed to Master.

**Pull request closed without merging**